torch/autograd/function.py

```diff
@@ -54,6 +55,65 @@ def save_for_backward(self, *tensors: torch.Tensor):
         """
         self.to_save = tensors
 
+    def save_for_forward(self, *tensors: torch.Tensor):
+        r"""Saves given tensors for a future call to :func:`~Function.forward`.
+
+        This should be only called from inside the :func:`forward` method.
+        This method should only be called with tensors. Unlike `save_for_backward`,
+        the passed tensors need not be an input or output.
+
+        In :func:`jvp`, saved objects can be accessed through the :attr:`saved_tensors`
+        attribute. Before returning them to the user, the primals of the dual tensors
+        are unpacked for convenience.
+
+        Arguments can also be ``None``. This is a no-op.
+
+        See :ref:`extending-autograd` for more details on how to use this method.
+
+        Example::
+            >>> class Func(torch.autograd.Function):
+            >>>     @staticmethod
+            >>>     def forward(ctx, x: torch.Tensor, y: torch.Tensor, z: int):
+            >>>         ctx.save_for_backward(x, y)
+            >>>         ctx.save_for_forward(x, y)
+            >>>         ctx.z = z
+            >>>         ctx.prod = x * y
+            >>>         return z * ctx.prod
+            >>>
+            >>>     @staticmethod
+            >>>     def jvp(ctx, x_t, y_t, _):
+            >>>         x_p, y_p = ctx.saved_tensors
+            >>>         z = ctx.z
+            >>>         return z * (y_p * x_t + x_p * y_t)
+            >>>
+            >>>     @staticmethod
+            >>>     def vjp(ctx, grad_out):
+            >>>         x, y = ctx.saved_tensors
+            >>>         z = ctx.z
+            >>>         return z * grad_out * y, z * grad_out * x, None
+            >>>
+            >>>     a = torch.tensor(1., requires_grad=True, dtype=torch.double)
+            >>>     t = torch.tensor(1., dtype=torch.double)
+            >>>     b = torch.tensor(2., requires_grad=True, dtype=torch.double)
+            >>>     c = 4
+            >>>
+            >>>     with fwAD.dual_level():
+            >>>         a_dual = fwAD.make_dual(a, t)
+            >>>         d = Func.apply(a_dual, b, c)
+
+        """
+        primals = []
+        for tensor in tensors:
+            assert isinstance(tensor, torch.Tensor) or tensor is None, ("save_for_forward"
+                "expects all arguments to be tensors; you should pass non-tensors as "
+                "attributes on ctx.")
+            if tensor is None:
+                primals.append(None)
+            else:
+                primals.append(fwAD.unpack_dual(tensor).primal)
```

Comment:
I think we should unpack dual here and only provide the primals in ctx.saved_tensor. All the necessary tangents for computation are already provided as arguments, and because forward AD is not run in forward, no intermediates have tangents.

Reply:
I also prefer this solution because unpacking may be confusing (there will be two copies of tangents), and sometimes if one of the inputs does not have grad, the undefined tangent will not be materialized for the unpacked dual tensor, but will be as it is passed as input to jvp.

Reply:
Without automatic unpacking, users forgetting to unpack would also run into the following error: `RuntimeError: Setting a forward grad that itself has a forward gradient at the same level0 is not supported.`

Reply:
Yes unpacking sounds good!